### ISSUE TITLE
With `GitPython`, getting false warning `"Repo" is not exported from module "git"`

### ISSUE THREAD

**Issue by DeflateAwning** [microsoft/pyright#5929]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

If you have a question about typing or a behavior that you’re seeing in Pyright (as opposed to a bug report or enhancement request), consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Describe the bug**
`pip install GitPython==3.1.35`

Make a file with:
```python
import git
repo = git.Repo(__file__)
```

See that there's an error reported from Pyright: `"Repo" is not exported from module "git"`

I believe it has to do with the way that `__all__` is dynamically generated in their lib's `__init__.py`. This is valid though.

**VS Code extension or command-line**
VS Code extension. Version 1.1.326.

**Comment by erictraut** [comment]
Pyright's behavior is correct here. `GitPython` is a "py.typed" library, which means it claims to contain type information and follow certain rules about indicating which symbols are public vs private. The `git/__init__.py` file contains the following import: `from git.repo import Repo`. By default, all imports in a py.typed library are considered private unless they are explicitly re-exported. To indicate that an imported symbol is intended to be re-exported, the maintainers of this library would need to use one of the techniques documented [here](https://microsoft.github.io/pyright/#/typed-libraries?id=library-interface). For example, if you manually change this import to `from git.repo import Repo as Repo` (using a redundant form), pyright and other type checkers will then interpret `Repo` as a public export from the top-level `git` module.

**Comment by DeflateAwning** [comment]
There is currently this line in the __init__.py file though: 

```python
__all__ = [name for name, obj in locals().items() if not (name.startswith("_") or inspect.ismodule(obj))]
```

Shouldn't this correctly make all public ones available?

**Comment by erictraut** [comment]
That's too dynamic for a static type checker to support. Keep in mind that static analysis tools don't actually execute the code. The maintainers of various static tools have agreed upon a subset of `__all__` expression forms that they will support. You can find the full list [here](https://microsoft.github.io/pyright/#/typed-libraries?id=library-interface). The form used in this library is not a supported form, and it's way too complicated for static analysis tools to handle. If the maintainers want their library to work with type checkers, language servers, linters, etc., they will need to switch to one of the supported forms.

**Comment by dimaqq** [comment]
It's a little odd that `pyright` gives off an Error that something is not exported form some module, yat, actually analyses if that something is called correctly.

This may be good for code from user's repo.

But it's hardly useful when something is imported from a 3rd party library.

Would it be reasonable to suppress errors for things from `site-packages`?
